### PR TITLE
Conveniences to make groups easier to configure in YAML

### PR DESCRIPTION
**Pull request by alasdairhodge**
* Extends `$brooklyn:object()` syntax to support parameterised constructors and static factory methods
* Adds variants of `EntityPredicates.config*()` and `attribute*()` that accept strings
* Adds variants of `EntityFunctions.config()` and `attribute()` that accept strings
* Adds variants of `EntityFunctions.config()` and `attribute()` that accept format specifier

### CODE REVIEW

File: camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/BrooklynDslCommon.java [hunk 1]
```diff
@@ -327,25 +337,74 @@ public FormatString(String pattern, Object[] args) {
...
+        private final String factoryMethodName;
+        private final Class<?> type;
+        private final List<Object> constructorArgs, factoryMethodArgs;
+        private final Map<String, Object> fields, config;
+
+        public DslObject(
+                String typeName,
+                List<Object> constructorArgs,
+                Map<String, Object> fields,
+                Map<String, Object> config) {
```
Comment (aledsage) [inline camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/BrooklynDslCommon.java]: I'd use `Map<String, ?>` if we can get away with it - makes it easier for the caller.
But this is pretty internal stuff, so I don't really care.

File: camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/BrooklynDslCommon.java [hunk 2]
```diff
@@ -384,24 +445,48 @@ public Object apply(List<Object> input) {
...
+            } catch (Exception e) {
+                throw Exceptions.propagate(e);
+            }
+        }
+
+        public static Object create(Class<?> type, String factoryMethodName, List<Object> factoryMethodArgs, Map<String,?> fields, Map<String,?> config) {
+            try {
+                Object bean;
+                try {
+                    bean = TypeCoercions.coerce(fields, type);
```
Comment (aledsage) [inline camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/BrooklynDslCommon.java]: I don't follow this line. Why are we trying to coerce the fields (which is a map) to the expected object type?

Do we expect to support someone using `$brooklyn:object: { type: Map, object.fields: { a: b } }` to instantiate a map with contents `{a: b}`?!

File: camp/camp-brooklyn/src/test/java/org/apache/brooklyn/camp/brooklyn/ObjectsYamlTest.java [hunk 3]
```diff
@@ -275,6 +288,57 @@ public void testBrooklynObjectWithFunction() throws Exception {
         Assert.assertTrue(testObjectObject instanceof SimpleTestPojo, "Expected a SimpleTestPojo: "+testObjectObject);
     }
 
+    @Test
+    public void testBrooklynObjectWithParameterisedConstructor() throws Exception {
```
Comment (aledsage) [inline camp/camp-brooklyn/src/test/java/org/apache/brooklyn/camp/brooklyn/ObjectsYamlTest.java]: Looks good. Would also be nice to have a test that used a deferred supplier. But given there are no existing tests in `ObjectsYamlTest` that do that, I see why you didn't either!

We can do that in a separate PR.

**Comment by aledsage** [comment]
@alasdairhodge looks good. Just one comment that I'm interested in before this is merged: what is the purpose of calling `TypeCoercions.coerce(fields, type)`?

File: camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/BrooklynDslCommon.java (hunk 2 above)
Comment (alasdairhodge) [inline camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/BrooklynDslCommon.java]: Yeah, I don't really know why that was in the previous version either: I can't think of a sensible use-case, and the unit tests don't cover it. I opted to preserve it rather than risk breaking some obscure corner case, but I'd be *very* happy to see it removed.

File: camp/camp-brooklyn/src/test/java/org/apache/brooklyn/camp/brooklyn/ObjectsYamlTest.java (hunk 3 above)
Comment (alasdairhodge) [inline camp/camp-brooklyn/src/test/java/org/apache/brooklyn/camp/brooklyn/ObjectsYamlTest.java]: +1

File: camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/BrooklynDslCommon.java (hunk 1 above)
Comment (alasdairhodge) [inline camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/BrooklynDslCommon.java]: I originally attempted this throughout, but it caused issues with the `Map.put` calls on lines 435, 437, 443, 445, as well as the call to `Reflections.invokeMethodFromArgs` on line 483.

**Comment by alasdairhodge** [comment]
Thanks, Aled. Suggest improving test coverage and tidying up the bizarre coersion stuff in a subsequent PR.

**Comment by alasdairhodge** [comment]
Just one other thing: do we have a naming convention for config keys? Don't want to start a religious war over the camelCase `factoryMethod` ;o)

**Comment by aledsage** [comment]
@alasdairhodge we don't have a consistent convention for config key names. We have three approaches being used:
* dot-separated - normally used as a hierarchy, but sometimes dots used in the same way as camelCase (but it shouldn't be)
* underscores
* camelCase

Agree would be nice to have consistency! I'm fine with your camelCase.

Merging now.
